Updates arrays section to improve order and clarify concepts (#5246)
ethereum/solidity: docs/types.rst

Arrays
------

Arrays can have a compile-time fixed size, or they can have a dynamic size.

An array of fixed size ``k`` and element type ``T`` is written as ``T[k]``,
and an array of dynamic size as ``T[]``.

For example, an array of 5 dynamic arrays of ``uint`` as ``uint[][5]``,

> the notation is reversed compared to some other languages.

Reviewer: I think we should spend some more time on this fact. We could add: The reason behind as that in Solidity, `X[3]` is always an array of three elements of type `X`, even if `X` is itself an array. This is not the case in e.g. C.

Indices are zero-based, and access is in the opposite direction of the declaration.

> For example to access the second uint in the third dynamic array, use ``x[2][1]``,

Reviewer: Please add the `x` also to the type in line 806, so it gets clear what `x` is.

> and to access the third dynamic array, use ``x[2]``

Reviewer: Add: Again, if you have an array `T[5] a` for some type `T` that can also be an array, then `a[2]` will always have type `T`.

Reviewer: `.` missing.

> Array elements can be of any type, including mapping or a struct. The general

Reviewer: why "a" struct but just mapping?

restrictions for types apply, in that mappings can only be stored in the
storage data location and publicly-visible functions need parameters that are :ref:`ABI types <ABI>`.

> Accessing an array past its end causes a revert, use the ``.push()`` method or

Reviewer: Perhaps clarify: not a revert but a failing assertion.

Arrays
------

Arrays can have a compile-time fixed size, or they can have a dynamic size.

> An array of fixed size ``k`` and element type ``T`` is written as ``T[k]``,

Reviewer: ```suggestion
The type of an array of fixed size ``k`` and element type ``T`` is written as ``T[k]``,
```

and an array of dynamic size as ``T[]``.

For example, an array of 5 dynamic arrays of ``uint`` is written as

> ``uint[][5]``. The notation is reversed compared to some other languages, as in

Reviewer: ```suggestion
``uint[][5]``. The notation is reversed compared to some other languages. In
```

Solidity, ``X[3]`` is always an array containing three elements of type ``X``,
even if ``X`` is itself an array. This is not the case in other languages such
as C.

Indices are zero-based, and access is in the opposite direction of the
declaration.

> For example to access the second ``uint`` in the third dynamic array, use

Reviewer: ```suggestion
For example if we have a variable `uint[][5] x memory`, you access the second ``uint`` in the third dynamic array using
```

Author: @chriseth Accepted, but with use of `you` in both cases.

Reviewer: Ok, but please still introduce `x` properly.

> ``unit[2][1]``, and to access the third dynamic array, use ``unit[2]``. Again,

Reviewer: ```suggestion
``x[2][1]``, and to access the third dynamic array, use ``x[2]``. Again,
```

if you have an array ``T[5] a`` for a type ``T`` that can also be an array,
then ``a[2]`` always has type ``T``.

Array elements can be of any type, including mapping or struct. The general

> restrictions for types apply, in that mappings can only be stored in the

Reviewer: "stored in storage"

Author: @chriseth Yeah, not too happy about it, but it's the word we've been using so far and I think consistency trumps the dodgy grammar. I code formatted ``storage`` to make it a little better.

Reviewer: My point is that "data location" is just the thing attached a type name. Here we also just call it "storage": https://solidity.readthedocs.io/en/v0.5.1/introduction-to-smart-contracts.html#storage-memory-and-the-stack

Author: @chriseth OK, I understand, you mean use 

"stored in storage" instead of "stored in the
``storage`` data location", I missed that from the comment.

OK, I am not opposed to either description, but at some point in the past few months we decided to standardise on "``x`` data location", but of course, this was in progress of change as relevant PRs were submitted. All I need to know is which one to stick with from now on and that's fine and I can correct references to the one style.

Reviewer: "In `StructName storage x` the variable `x` has a data location of `storage` and thus its data resides in storage."

storage data location and publicly-visible functions need parameters that are :ref:`ABI types <ABI>`.

> Accessing an array past its end causes a failing assertion, use the ``.push()``

Reviewer: ```suggestion
Accessing an array past its end causes a failing assertion. You can use the ``.push()`` method to append a new element at the end or assign to the ``.length`` :ref:`member <array-members>` to change the size (see below for caveats).
```

method or increase the ``.length`` :ref:`member <array-members>` to add elements.

Variables of type ``bytes`` and ``string`` are special arrays. A ``bytes`` is similar to ``byte[]``,
but it is packed tightly in calldata and memory. ``string`` is equal to ``bytes`` but does not allow
length or index access.

> You should use ``bytes`` over ``byte[]`` because it is cheaper. As a general rule,

Reviewer: ```suggestion
You should use ``bytes`` over ``byte[]`` because it is cheaper, since ``byte[]`` adds 31 padding bytes between the elements. As a general rule,
```

use ``bytes`` for arbitrary-length raw byte data and ``string`` for arbitrary-length
string (UTF-8) data. If you can limit the length to a certain number of bytes,

> always use one of ``bytes1`` to ``bytes32`` because they are much cheaper.

Reviewer: ```suggestion
always use one of the value types ``bytes1`` to ``bytes32`` because they are much cheaper.
```

.. index:: ! array;length, length, push, pop, !array;push, !array;pop

> .. _members:

Reviewer: Please name this more specifically, it is just about array members.
